### PR TITLE
General refactor

### PR DESCRIPTION
**Pull request by iulianpascalau**
- complete refactor of the solution: split functionality in packages and structures, added unit/semi-integration unit tests

### CODE REVIEW

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: checking/errors.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import "errors"
+
+// ErrNilValue signals that a nil value was provided
+var ErrNilValue = errors.New("nil value")
+
+// ErrZeroOrNegative signals that a zero or negative value was provided
+var ErrZeroOrNegative = errors.New("zero or negative value")
+
+// ErrEmptyInitialAccounts signals that an empty initial accounts list
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply checking/errors.go:11]: done

File: checking/errors.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+var ErrEmptyInitialAccounts = errors.New("empty initial accounts list")
+
+// ErrSupplyMismatch signals that the initial account's supply mismatches
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply checking/errors.go:14]: rephrased

File: checking/errors.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+// ErrStakingValueError signals the provided staking value is not a multiple of the node's price
+var ErrStakingValueError = errors.New("staking value error")
+
+// ErrNegativeValue signals that the provided value is negative
+var ErrNegativeValue = errors.New("negative value")
+
+// ErrDelegationValues signals that the provided delegation fields were improperly set
+var ErrDelegationValues = errors.New("delegation values error")
+
+// ErrTotalSupplyMismatch signals that the total supply mismatches
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply checking/errors.go:26]: rephrased

File: cmd/filegen/main.go [hunk 4]
```diff
@@ -76,17 +64,17 @@ VERSION:
 	numOfShards = cli.IntFlag{
 		Name:  "num-of-shards",
 		Usage: "Number of initial shards",
-		Value: 5,
+		Value: 3,
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/filegen/main.go:67]: I would keep them as they are as we won't test for now with 5 shards. Will update the readme.md also

File: core/converters.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+// ConvertToBigInt will try to convert the provided string to its big int corresponding value
+func ConvertToBigInt(value string) (*big.Int, error) {
+	valueNumber, isNumber := big.NewInt(0).SetString(value, 10)
+	if !isNumber {
+		return nil, ErrStringIsNotANumber
+	}
+
+	if valueNumber.Cmp(big.NewInt(0)) < 0 {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply core/converters.go:21]: Agree, changed the function name to `ConvertToPositiveBigInt`

File: core/converters.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// GenerateSCAddress will generate the resulting SC address from the provided public key string and nonce
+func GenerateSCAddress(
+	pkString string,
+	nonce uint64,
+	vmType string,
+	converter elrondCore.PubkeyConverter,
+) (string, error) {
+
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply core/converters.go:35]: done

File: core/converters_test.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+	pkConv, _ := pubkeyConverter.NewBech32PubkeyConverter(32)
+	scAddress, err := GenerateSCAddress(
+		"erd1ulhw20j7jvgfgak5p05kv667k5k9f320sgef5ayxkt9784ql0zssrzyhjp",
+		0,
+		"0500",
+		pkConv,
+	)
+
+	require.Nil(t, err)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply core/converters_test.go:48]: no necessarily true. We can have both. Last checks can be asserts if they can't panic. I have followed the rule: if the test can panic in an assert (due to nils or whatever), the preceding check(s) should be done with require. Having as many as asserts we can put will throw all errors at once. The use of require will stop at each error (the test will stop there). In some cases it is better to see all errors (or as much errors as we can get) in order to easily debug the code.

File: core/fileHandler.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	return pem.Encode(fh, &blk)
+}
+
+// Close will try to close the file
+func (fh *fileHandler) Close() {
+	err := fh.File.Close()
+	log.LogIfError(err)
+}
+
+// IsInterfaceNil returns if underlying object is true
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply core/fileHandler.go:65]: fixed all of them :+1:

File: generating/delegatedGenerator.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+			Address:       dsg.delegationScPkString,
+			InitialRating: dsg.initialRating,
+		}
+		initialNodes = append(initialNodes, initialNode)
+	}
+
+	return initialNodes
+}
+
+// IsInterfaceNil returns if underlying object is true
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply generating/delegatedGenerator.go:229]: done

File: generating/delegatedGenerator_test.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+	arg.NodePrice.SetString("2500000000000000000000", 10)
+
+	return arg
+}
+
+func TestDelegatedStakingGenerator_GenerateShouldWork(t *testing.T) {
+	t.Parallel()
+
+	arg := createMockDelegatedStakingGeneratorArguments()
+	arg.NumValidatorBlsKeys = 33
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply generating/delegatedGenerator_test.go:47]: used args fields as variables

File: generating/directStakingGenerator.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+			Address:       dsg.walletPubKeyConverter.Encode(key.PubKeyBytes),
+			InitialRating: dsg.initialRating,
+		}
+		initialNodes = append(initialNodes, initialNode)
+	}
+
+	return initialNodes
+}
+
+// IsInterfaceNil returns if underlying object is true
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply generating/directStakingGenerator.go:183]: done

File: generating/directStakingGenerator_test.go [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+	arg.ValidatorPubKeyConverter, _ = pubkeyConverter.NewHexPubkeyConverter(96)
+
+	return arg
+}
+
+func TestDirectStakingGenerator_GenerateShouldWork(t *testing.T) {
+	t.Parallel()
+
+	arg := createMockDirectStakingGeneratorArguments()
+	arg.NumValidatorBlsKeys = 33
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply generating/directStakingGenerator_test.go:44]: done

File: generating/disabled/nilRandomizer.go [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+// NilRandomizer is a mock implementation of the IntRandomizer interface
+type NilRandomizer struct {
+}
+
+// Intn returns 0
+func (nr *NilRandomizer) Intn(_ int) int {
+	return 0
+}
+
+// IsInterfaceNil returns if underlying object is true
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply generating/disabled/nilRandomizer.go:12]: done

File: generating/validatorKeyGenerator.go [hunk 14]
```diff
@@ -0,0 +1,49 @@
+package generating
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply generating/validatorKeyGenerator.go:1]: renamed: `generate` and `check`

File: io/errors.go [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package io
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply io/errors.go:1]: renamed to plugins.

File: io/outputHandler.go [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+		Adaptivity:                  oh.adaptivityValue,
+		ChainID:                     oh.chainID,
+		MinTransactionVersion:       uint32(oh.txVersion),
+		InitialNodes:                initialNodes,
+	}
+
+	return oh.nodesSetupHandler.WriteObjectInFile(nodes)
+}
+
+// writeGenesisFile will write the provided initial accounts in to the genesis file
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply io/outputHandler.go:128]: removed

File: io/outputHandler.go [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+	oh.nodesSetupHandler.Close()
+	if !check.IfNil(oh.txgenAccountsHandler) {
+		oh.txgenAccountsHandler.Close()
+	}
+	if !check.IfNil(oh.delegatorsHandler) {
+		oh.delegatorsHandler.Close()
+	}
+}
+
+// IsInterfaceNil returns if underlying object is true
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply io/outputHandler.go:260]: done

File: io/txgenAccount.go [hunk 18]
```diff
@@ -0,0 +1,12 @@
+package io
+
+import "math/big"
+
+type txgenAccount struct {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply io/txgenAccount.go:5]: Will move it if we need it in some other places. This struct is used only in the plugins package and it is not a business logic core data structure.

**Review by bogdan-rosianu — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Pull request merged**